File: python/bot/cogs/moderation/warn.py
"""?warn — warning system."""

from __future__ import annotations

import logging

import discord
from discord.ext import commands

from bot.checks.decorators import moderator_required
from bot.cogs.deps import CogRepos
from bot.services.hierarchy import get_moderation_denied
from bot.services.mod_log import send_mod_log
from bot.services.strikes import check_strike_escalation
from bot.utils.helpers import basic_embed, error, success
from bot.utils.resolvers import resolve_member
from bot.utils.time import format_iso_date

logger = logging.getLogger(__name__)


class WarnCog(commands.Cog):
    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot
        self.repos = CogRepos(bot)

    @commands.group(name="warn", invoke_without_command=True)
    @moderator_required()
    async def warn(self, ctx: commands.Context[commands.Bot]) -> None:
        await ctx.reply(error("Usage: `?warn add|list|del|clear|view|remove`"))

    @warn.command(name="add")
    @moderator_required()
    async def warn_add(self, ctx: commands.Context[commands.Bot], *, args: str) -> None:
        parts = args.split()
        target = resolve_member(ctx.guild, ctx.message, parts[0])  # type: ignore[arg-type]
        if target is None:
            await ctx.reply(error("That user is not currently a member of this server."))
            return
        if not isinstance(ctx.author, discord.Member):
            return
        denied = get_moderation_denied(ctx.guild, ctx.author, target)  # type: ignore[arg-type]
        if denied:
            await ctx.reply(denied)
            return
        reason = " ".join(parts[1:]) or "No reason provided"
        try:
            warning_id, case_number = await self.repos.warnings.create_with_case(
                guild_id=str(ctx.guild.id),  # type: ignore[union-attr]
                user_id=str(target.id),
                moderator_id=str(ctx.author.id),
                reason=reason,
                source="warn_command",
                cases=self.repos.cases,
            )
        except Exception as exc:
            logger.error("[warn] Database write failed: %s", exc)
            await ctx.reply(error("Could not save that warning."))
            return

        notified = await send_mod_log(
            ctx.guild,  # type: ignore[arg-type]
            action="warn",
            target=target,
            moderator=ctx.author,
            reason=reason,
            case_number=case_number,
            guild_settings=self.repos.guild_settings,
        )
        if not notified:
            logger.error("[warn] Mod-log channel notification failed for case #%s", case_number)

        reply = success(
            f"Warned **{target.display_name}** — Warning #{warning_id}, Case #{case_number}. Reason: {reason}",
        )
        escalation = await check_strike_escalation(
            ctx.guild,  # type: ignore[arg-type]
            target,
            ctx.author,
            guild_settings=self.repos.guild_settings,
            warnings=self.repos.warnings,
            cases=self.repos.cases,
            strike_state=self.repos.strike_state,
        )
        if escalation:
            extra = escalation.value if escalation.ok else escalation.error
            if extra:
                reply = f"{reply}\n{extra}"
        await ctx.reply(reply)

    @warn.command(name="list", aliases=["view"])
    @moderator_required()
    async def warn_list(self, ctx: commands.Context[commands.Bot], *, user_arg: str | None = None) -> None:
        target = resolve_member(ctx.guild, ctx.message, user_arg)  # type: ignore[arg-type]
        if target is None and isinstance(ctx.author, discord.Member):
            target = ctx.author
        if target is None:
            await ctx.reply(error("Usage: `?warn list [@user]`"))
            return
        warnings = await self.repos.warnings.list_active(str(ctx.guild.id), str(target.id))  # type: ignore[union-attr]
        if not warnings:
            await ctx.reply(error(f"**{target.display_name}** has no warnings."))
            return
        lines = [f"**#{w.id}** — {w.reason or 'No reason'} ({format_iso_date(w.created_at)})" for w in warnings]
        await ctx.reply(
            embed=basic_embed(f"Warnings: {target.display_name}", "\n".join(lines), color=0xFEE75C),
        )

    @warn.command(name="del", aliases=["remove"])
    @moderator_required()
    async def warn_del(self, ctx: commands.Context[commands.Bot], warning_id: str) -> None:
        parsed = int(warning_id.replace("#", ""), 10)
        if not parsed:
            await ctx.reply(error("Usage: `?warn del <warning ID>`"))
            return
        if not await self.repos.warnings.void(parsed):
            await ctx.reply(error("Warning not found."))
            return
        await ctx.reply(success(f"Deleted warning #{parsed}."))

    @warn.command(name="clear")
    @moderator_required()
    async def warn_clear(self, ctx: commands.Context[commands.Bot], *, user_arg: str) -> None:
        target = resolve_member(ctx.guild, ctx.message, user_arg)  # type: ignore[arg-type]
        if target is None:
            await ctx.reply(error("Usage: `?warn clear <user>`"))
            return
        await self.repos.warnings.void_all_for_user(str(ctx.guild.id), str(target.id))  # type: ignore[union-attr]
        await ctx.reply(success(f"Cleared all warnings for **{target.display_name}**."))


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(WarnCog(bot))

File: node-bun/src/commands/moderation/warn.js
import { getModerationDenied } from '../../utils/permissions.js';
import { resolveMember, success, error, basicEmbed } from '../../utils/helpers.js';
import { checkMod } from '../../utils/checks.js';
import { sendModLog } from '../../utils/modLog.js';
import { createWarningWithCase, getWarnings, deleteWarning, clearWarnings } from '../../database/db.js';
import { checkStrikeEscalation } from '../../utils/strikes.js';
import { formatDate } from '../../utils/time.js';

export default {
  name: 'warn',
  description: 'Warning system: add, list, del, clear',
  category: 'Moderation',
  usage: '?warn add|list|del|clear [args]',
  async execute(message, args, subcommand, subargs) {
    const denied = await checkMod(message);
    if (denied) return message.reply(denied);

    const action = subcommand?.toLowerCase() || 'add';
    const rest = subargs || args.trim();

    switch (action) {
      case 'add': {
        const parts = rest.split(/\s+/);
        const target = resolveMember(message, parts[0]);
        if (!target) return message.reply(error('That user is not currently a member of this server.'));
        const hierarchyDenied = getModerationDenied(message.guild, message.member, target);
        if (hierarchyDenied) return message.reply(hierarchyDenied);
        const reason = parts.slice(1).join(' ') || 'No reason provided';

        let records;
        try {
          records = await createWarningWithCase({
            guildId: message.guild.id,
            userId: target.id,
            moderatorId: message.author.id,
            reason,
            source: 'warn_command',
          });
        } catch (err) {
          console.error('[warn] Database write failed:', err.message);
          return message.reply(error('Could not save that warning.'));
        }

        const { warningId, caseNumber } = records;
        const notified = await sendModLog(message.guild, {
          action: 'warn',
          target,
          moderator: message.author,
          reason,
          caseNumber,
        });
        if (!notified) {
          console.error(`[warn] Mod-log channel notification failed for case #${caseNumber}`);
        }

        let reply = success(`Warned **${target.user.tag}** — Warning #${warningId}, Case #${caseNumber}. Reason: ${reason}`);
        const escalation = await checkStrikeEscalation(message.guild, target, message.author);
        if (escalation) reply += `\n${escalation}`;
        return message.reply(reply);
      }
      case 'list': {
        const target = resolveMember(message, rest.trim()) ?? message.member;
        const warnings = await getWarnings(message.guild.id, target.id);
        if (!warnings.length) return message.reply(error(`**${target.user.tag}** has no warnings.`));
        const lines = warnings.map((w) => `**#${w.id}** — ${w.reason || 'No reason'} (${formatDate(w.created_at)})`);
        return message.reply({ embeds: [basicEmbed(`Warnings: ${target.user.tag}`, lines.join('\n'), 0xfee75c)] });
      }
      case 'del': {
        const id = parseInt(rest.trim().replace('#', ''), 10);
        if (!id) return message.reply(error('Usage: `?warn del <warning ID>`'));
        const result = await deleteWarning(id);
        if (!result.changes) return message.reply(error('Warning not found.'));
        return message.reply(success(`Deleted warning #${id}.`));
      }
      case 'clear': {
        const target = resolveMember(message, rest.trim());
        if (!target) return message.reply(error('Usage: `?warn clear <user>`'));
        await clearWarnings(message.guild.id, target.id);
        return message.reply(success(`Cleared all warnings for **${target.user.tag}**.`));
      }
      default:
        return message.reply(error('Usage: `?warn add|list|del|clear`'));
    }
  },
};
